eq-author/src/components/PositionModal/index.js

```diff
@@ -32,64 +32,126 @@ const Trigger = styled.button.attrs({ type: "button" })`
+  const positionButtonId = uniqueId("PositionModal");
+  const [isOpen, setIsOpen] = useState(false);
   const previousPosition = useMemo(
     () => options.findIndex(({ id }) => id === selected.id),
     [options, selected]
   );
-  const [isOpen, setIsOpen] = useState(false);
-  const [selectedPosition, setSelectedPosition] = useState(previousPosition);
-  const [positionButtonId] = useState(uniqueId("PositionModal"));
+
+  const [{ position, item, isSame }, setOption] = useState({
+    position: previousPosition,
+    item: options[previousPosition],
+  });
 
   const orderedOptions = options.filter(({ id }) => id !== selected.id);
-  orderedOptions.splice(selectedPosition, 0, selected);
+  selected.parentEnabled = item?.parentEnabled;
+  orderedOptions.splice(position, 0, selected);
 
   const handleClose = () => {
     setIsOpen(false);
-    setSelectedPosition(previousPosition);
+    setOption({
+      position: previousPosition,
+      item: orderedOptions[previousPosition],
+    });
   };
 
-  const handleOpen = () => setIsOpen(true);
-
-  const handleChange = ({ value }) => setSelectedPosition(parseInt(value, 10));
+  const handleChange = ({ value }) =>
+    setOption(({ item: previousItem }) => {
+      const count =
+        orderedOptions[value].type === "Folder" && value - position >= 0
+          ? orderedOptions.filter(
+              ({ parentId }) => parentId === orderedOptions[value].id
+            ).length
+          : 0;
+      return {
+        position: parseInt(value, 10) + count,
+        item: orderedOptions[value],
+        isSame: previousItem?.id === orderedOptions[value]?.id,
+      };
+    });
 
   const handleConfirm = (e) => {
     e.preventDefault();
-    setIsOpen(false);
+    const { parentId, position: itemPosition, parentEnabled } = item;
+    const lastIndex = orderedOptions.findIndex(({ id }) => item.id === id);
+
+    const checkIndex = position - lastIndex > -1;
+
+    const differentSection =
+      selected?.section?.id !== orderedOptions[position]?.section?.id;
+
+    const intoFolder =
+      !orderedOptions[previousPosition]?.parentEnabled && parentEnabled;
+    const outOfFolder =
+      orderedOptions[previousPosition]?.parentEnabled && !parentEnabled;
+    const fromFolderToFolder =
+      orderedOptions[previousPosition]?.parentEnabled && parentEnabled;
+
+    const inOrOut = intoFolder || outOfFolder;
+
+    const sameMove = checkIndex ? itemPosition + 1 : itemPosition - 1;
+    const folderMove = checkIndex ? itemPosition + 1 : itemPosition;
+
+    let positionCalculation;
+
+    if (isSame) {
+      positionCalculation = sameMove > 0 ? sameMove : 0;
+    } else if (differentSection || fromFolderToFolder || inOrOut) {
+      positionCalculation = folderMove;
+    } else {
+      const selectedItemPosition = orderedOptions.findIndex(
+        ({ id }) => id === selected.id
+      );
+      positionCalculation = selectedItemPosition;
+    }
+
```

Comment:
Could you add a couple wee comments to this bit to explain what the logic behind all this arithmetic is? Could save Shane a headache when it comes to adding in moving folders ;-)

Reply:
Will do, I mean it looks self explanatory to me... ;)

Reply:
this i solved this